torchdata/nodes/batch.py

```diff
@@ -56,3 +56,56 @@ def next(self) -> List[T]:
 
     def get_state(self) -> Dict[str, Any]:
         return {self.SOURCE_KEY: self.source.state_dict()}
+
+
+class Unbatcher(BaseNode[T]):
+    """Unbatcher will flatten batches pulled from source, and
+    yields elements in sequential order when next() is called on it.
+
+    Args:
+        source (BaseNode[T]): The source node to pull batches from.
+    """
+
+    SOURCE_KEY = "source"
+    BATCH_IDX_KEY = "batch_idx"
+
+    def __init__(self, source: BaseNode[Sequence[T]]):
+        super().__init__(self)
+        self.source = source
+
+    def reset(self, initial_state: Optional[Dict[str, Any]] = None):
+        super().reset(initial_state)
+        if initial_state is not None:
+            self.source.reset(initial_state[self.SOURCE_KEY])
+            self._cached_state_dict = initial_state[self.SOURCE_KEY]
+            try:
+                self._batch = next(self.source)
+                self._batch_idx = initial_state[self.BATCH_IDX_KEY]
+            except StopIteration:
+                # next(self.source) will be called upon subsequent self.next() call
+                # and raise StopIteration in the correct place.
+                self._batch = []
+                self._batch_idx = 0
+        else:
+            self.source.reset()
+            self._batch = []
+            self._cached_state_dict = None
+            self._batch_idx = 0
+
+    def next(self) -> T:
+        while self._batch_idx >= len(self._batch):
```

Comment:
This should be `if` instead of `while` ? i.e. if the _batch_idx overshoots the current _batch, get a new _batch and reset _batch_idx to 0.

Reply:
@divyanshk yes i was worried about next batch of size 0 case, although I'm assuming that's unlikely but it's an edge case none the less